### DISCUSSION TITLE
design review for Jan 26

### DISCUSSION THREAD

**Discussion by randytpierce** [NOAA-GSL/MATS#856] [Scorecard - development]
The primary goal of this review is to discuss the approaches for a few things.
1) The mechanism for the scorecard display to open a new window to a specific mats application and a specific preconfigured plot that corresponds to the significance value in the scorecard (table cell) that the user clicked on.

2) The mechanism for the processScorecard backend matsMethod to calculate all the required significance values and populate the scorecard couchbase data record. The assumption is that this will be a very parallel worker set.

There are preliminary UML sequence diagrams in the github scorecard_display branch in the apps/scorecard/model folder. There is a README.md that explains how to use plantUml (which was used to generate the sequence diagrams).

**Comment by randytpierce** [comment]
Here is the readme about plantUML 
https://github.com/NOAA-GSL/MATS/blob/scorecard_display/apps/scorecard/model/README.md

**Comment by randytpierce** [comment]
matsPlot sequence

![image](https://user-images.githubusercontent.com/25647166/214717188-58f1a1a2-9d47-415a-899f-2b8a38633c55.png)

process data sequence

![image](https://user-images.githubusercontent.com/25647166/214717483-c5cefce8-e0b4-4bee-bb53-8f31c24b9c51.png)

**Comment by randytpierce** [comment]
Keeping a few notes here:
We will probably want to expire the scorecard mongo records. When a request is made to display a scorecard the scorecard is retrieved from Couchbase and inserted into a mongo collection that is reactive with the scorecard display templates. We wont want to keep all the scorecards in mongo so we should give them a time to live. See https://www.mongodb.com/docs/manual/tutorial/expire-data/ for how TTL works with mongo. What the time to live should be is a design question.

**Comment by randytpierce** [comment]
New scPlot Sequence

![image](https://user-images.githubusercontent.com/25647166/214909991-9921ebb4-bd28-4622-9dd5-6710cd51d1ed.png)

**Comment by randytpierce** [comment]
Rather than putting images in the discussion it is much better to go to 
https://github.com/NOAA-GSL/MATS/tree/scorecard_display/apps/scorecard/model
and copy a .puml file of interest and paste it into the online plantuml server. https://www.plantuml.com/plantuml/uml/SyfFKj2rKt3CoKnELR1Io4ZDoSa70000
Or go to https://plantuml.com/ and click on "online server" in the left column and then paste the contents of the .puml file into the text area and let it render. It automatically updates any changes. So these images are likely out of date already.

**Comment by randytpierce** [comment]
This is a potential list of issues, some of these will need subtasks. I am not creating the actual task list until after we review it.
ProcessScorecardData:

- implement scheduler (probable Bree?) in nodejs in MATS backend. 
- implement mechanism for scorecardApp.server.processScorecard to schedule a scorecard process
- design and create a calculate service. This should be a web enabled micro-service (I think in golang).
- Within calculate service we need to design and implement a  Builder pattern - see https://refactoring.guru/design-patterns/builder/go/example
- implement director and process queue
- implement scorecard builders as Go thread workers
- determine do we use AWS lambdas or do we use internal server? How do we do this? BTW - Lambda comes from the [Lambda Calculus](http://en.wikipedia.org/wiki/Lambda_calculus) and refers to anonymous functions in programming. Lambdas are possible in any language. AWS lambdas are anonymous functions that are automatically associated with scalable compute time. Theoretically, depending on Scotts budget we could scale AWS lambdas to a number that could finish off a scorecard very quickly - provided we can the Lamda functions the necessary data. So... question, "Do we do the data retrieval in our local director and pass the data to the Lambdas (this MIGHT seriously disadvantage the multithreaded nature of the whole thing) or do we migrate data to someplace that the Lamdas can retrieve it and pass the constructed queries (and database credentials) to the Lamdas?". This is a big question. 
- implement calculations - AWS Lamdas or go routines
- implement director's Couchbase scorecard update mecahnism.

scMatsPlot:
- Create the mongo collection (scorecards), subscription, etc. for the temporary scorecard  mongo documents
- Make the matsMethod.getScorecardData insert the ret data as a mongo record into mongo scorecards collection. Consider Time to Live and make necessary implementation to set TTL on the record.  See https://www.mongodb.com/docs/manual/tutorial/expire-data/ for how TTL works with mongo. 
- Rework (slightly) the scorecardDisplay to use the scorecard mongo collection
- Make the user account settings persistent across restarts and available to all server instances (pay attention to PI here).
- rework save/restore settings to be user oriented and persistent across restarts
- implement the creating of a settings (app plotParams, I believe) object from a scorecard plotParams
- implement the API for saving the settings to a server. (I sure would like to have these shared between instances somehow - store them in couchbase?) Make the scorecardDisplay do this.
- create the FlowRouter route for rendering an application from a "saved settings".

**Reply by ian-noaa** [reply]
Given that we want to persist and share these scorecards between users, it seems like the scorecard-related stuff should go in its own long-lived mongo database or in couchbase. We'll need to back these settings/configurations up in case something happens to the database. (Especially given that the current Mongo database has to be regularly purged) Having them in a long-lived database would also make it easier to update the scorecards in the background so that everything is up-to-date when users view the scorecard.

**Reply by randytpierce** [reply]
Absolutely. The scorecards themselves (the thing that holds the params and the data) is a document in the couchbase. Just retrieve the records i.e. "select * from vxdata._default.SCORECARD" (This only works against [adb-cb2.gsd.esrl.noaa.gov](https://adb-cb2.gsd.esrl.noaa.gov:18091/) or adb-cb3... or adb-cb4....) because the SCORECARDS only live on the cluster. 

The two entities that I don't yet have a good feel for are the settings (which won't be specific to scorecard) and the user account data. User account passwords are encrypted and I am wondering if I save the account data documents and reload them on another server do they still work?  Will the bcrypt decryption still work. I doubt it. So I am wondering if we have to use login.gov and offload the distribution of password problem to them.

**Comment by randytpierce** [comment]
An example of a significance calculation. It demonstrates the complexity of the calculation. I haven't verified this, I just borrowed it from some code.

```js
// use a student's t-test to see if a point on a contour diff is statistically significant
// see https://en.wikipedia.org/wiki/Student's_t-test
// The t-test's most common application is to test whether the means of two populations are different.
// x1,x2, are means of two populations, s1,s2, are standard deviations, and n1, and n2 are the number of individuals in the populations.
// We would possibly offload the means calculations to the database query and optimize the query with indexing.
function (x1, x2, s1, s2, n1, n2, sigType) {
    const t = Math.abs(x1 - x2) / Math.sqrt((Math.pow(s1, 2) / n1) + (Math.pow(s2, 2) / n2));
    const df = Math.pow(((Math.pow(s1, 2) / n1) + (Math.pow(s2, 2) / n2)), 2) / ((1 / (n1 - 1)) * Math.pow((Math.pow(s1, 2) / n1), 2) + (1 / (n2 - 1)) * Math.pow((Math.pow(s2, 2) / n2), 2));
    return isStudentTTestValueSignificant(t, df, sigType);
};

const isStudentTTestValueSignificant = function (t, df, sigType) {
    const sigThreshs = {
        1: 12.706,
        2: 4.303,
        3: 3.182,
        4: 2.776,
        5: 2.571,
        6: 2.477,
        7: 2.365,
        8: 2.306,
        9: 2.262,
        10: 2.228,
        11: 2.201,
        12: 2.179,
        13: 2.160,
        14: 2.145,
        15: 2.131,
        16: 2.120,
        17: 2.110,
        18: 2.101,
        19: 2.093,
        20: 2.086,
        21: 2.080,
        22: 2.074,
        23: 2.069,
        24: 2.064,
        25: 2.060,
        26: 2.056,
        27: 2.052,
        28: 2.048,
        29: 2.043,
        30: 2.042
    };

    var sigThresh;
    if (sigType === 'standard') {
        if (df <= 30) {
            sigThresh = sigThreshs[df];
        } else if (df <= 40) {
            sigThresh = 2.021;
        } else if (df <= 60) {
            sigThresh = 2.000
        } else if (df <= 120) {
            sigThresh = 1.980
        } else {
            sigThresh = 1.960
        }
    } else {
        sigThresh = 1.960
    }

    return t > sigThresh
};
```

**Comment by mollybsmith-noaa** [comment]
Comment: Make text cell backgrounds white.

**Reply by randytpierce** [reply]
sure
Just the cell background?

**Comment by mollybsmith-noaa** [comment]
Comment: Ceiling and visibility do need separate metadata, so instead of changing that there, we need a dictionary to interpret app names or something.

**Reply by randytpierce** [reply]
This is a sticky wicket, but it must be played.

**Comment by ian-noaa** [comment]
## AWS Lambda vs an on-prem system

_Follow up from the 1/26 meeting_

Given that this has to be working in some fashion by April, we decided that AWS Lambdas are probably not possible for the initial DDRF. However, if we were to use Lambdas we would most likely want to make heavy use of a message queue. It'd be good to keep in mind how we could refactor the backend's processes to fit that model.

**Reply by randytpierce** [reply]
I like the idea of a message queue. I think that we have to consider that the director and its workers might not be colocated.

**Comment by ian-noaa** [comment]
## Project Scope

_Follow up from the 1/26 meeting_

Matt had a great question: Given that this needs to be done by April - are there ways to reduce the scope or to make the implementation easier? E.g. - can we only process certain variables or can we limit the flexibility of the application (by say only processing set intervals) to make our lives easier?

**Reply by randytpierce** [reply]
I spent quite a bit of time thinking about this. The primary functionality is repeatable, I mean if it works for one variable it will work on 10. So limiting the datasources, regions, variables, or levels may not help. Limiting the apps might help. And we may have to say something like "if you make a huge scorecard it might take awhile".

**Comment by ian-noaa** [comment]
## Scorecard updating

_Follow up from the 1/26 meeting_

It was unclear to me how scorecard updates were being scheduled. Is this something that will be triggered on a schedule by the backend? Or will these calculations be triggered when a user loads the scorecard page? Or something else? The first seems preferable unless someone wants a one-off scorecard.

**Reply by randytpierce** [reply]
Do you mean updates from the CalculateService? That happens when the director process gets something from the workers worth updating. By worth updating I mean when enough has been done to be useful but not so often that it hamstrings the database. The update method is likely a server side route that updates the scorecard in couchbase and the mongo record, If it exists.

**Comment by ian-noaa** [comment]
## Scheduling

_Follow up from the 1/26 meeting_

Scheduled/batched operations tend to happen at the same time. E.g. - cron jobs tend to run at the top of the hour, etc... When the scheduler makes a large number of requests how will we ensure the requests aren't lost? This seems like a great place for a message queue. Otherwise, the backend could queue up the requests to hand off to the workers. (Until it ran out of memory)

If we didn't want to do a queue, we could start by having multiple backends running in the lab's Kubernetes cluster and using a Kubernetes service/load balancer to distribute requests between them.

**Reply by randytpierce** [reply]
Correct, except that the director has to know it's resource limits. For the ingest code that limit is number of threads. I suspect that for this it won't be memory, it will also be the number of allowed workers that can simultaneously run. Each worker does this...
1) know its lambda function (actually the director has to associate the proper data and lambda function with the worker)
2) pull an item off the queue (maybe we have multiple queues that are associated with specific lambda functions?
3) process the item
4) return the result
and then start over at 1) until there are no more items left on the queue. Then it dies. When they are all dead the director probably dies and the service will start another director when a new request comes in. The director starts the workers.

**Comment by ian-noaa** [comment]
## Persisting Schedules

_Follow up from the 1/26 meeting_

Where does the NPM scheduler persist its schedules? (I.e. - what happens when we destroy and recreate the app and mongo containers)

**Reply by randytpierce** [reply]
I don't remember where Bree has its persistence. I think there are options for it. I am thinking it is file based, actually. I know it persists over restarts.

**Reply by ian-noaa** [reply]
Gotcha. If there are options to persist it out to couchbase or an external store somewhere, that'd be helpful. GSL's Kubernetes cluster has PVC's but I'd be a little less comfortable relying on them for stuff that can't be easily recreated. (Like our current usage of Mongo)

**Comment by mollybsmith-noaa** [comment]
Anything that has text? It’s hard to read on the textured background.

**Reply by randytpierce** [reply]
I defer to your judgement. Feel free to spend as much time as you like on that stylesheet!
